python/dgl/nn/pytorch/explain/pgexplainer.py

```diff
@@ -348,13 +346,225 @@ def explain_graph(self, graph, feat, tmp=1.0, training=False, **kwargs):
+        >>> class Model(nn.Module):
+        ...     def __init__(self, in_feats, hid_feats, out_feats, rel_names):
+        ...         super().__init__()
+        ...         self.conv = dgl.nn.HeteroGraphConv(
+        ...             {rel: dgl.nn.GraphConv(in_feats, hid_feats) for rel in rel_names},
+        ...             aggregate="sum",
+        ...         )
+        ...         self.fc = nn.Linear(hid_feats, out_feats)
+        ...         nn.init.xavier_uniform_(self.fc.weight)
+        ...
+        ...     def forward(self, g, h, embed=False, edge_weight=None):
+        ...         if edge_weight:
+        ...             mod_kwargs = {
+        ...                 etype: {"edge_weight": mask} for etype, mask in edge_weight.items()
+        ...             }
+        ...             h = self.conv(g, h, mod_kwargs=mod_kwargs)
+        ...         else:
+        ...             h = self.conv(g, h)
+        ...
+        ...         if embed:
+        ...             return h
+        ...
+        ...         with g.local_scope():
+        ...             g.ndata["h"] = h
+        ...             hg = 0
+        ...             for ntype in g.ntypes:
+        ...                 hg = hg + dgl.mean_nodes(g, "h", ntype=ntype)
+        ...             return self.fc(hg)
+
+        >>> # Load dataset
+        >>> input_dim = 5
+        >>> hidden_dim = 5
+        >>> num_classes = 2
+        >>> g = dgl.heterograph({("user", "plays", "game"): ([0, 1, 1, 2], [0, 0, 1, 1])})
+        >>> g.nodes["user"].data["h"] = th.randn(g.num_nodes("user"), input_dim)
+        >>> g.nodes["game"].data["h"] = th.randn(g.num_nodes("game"), input_dim)
+
+        >>> transform = dgl.transforms.AddReverse()
+        >>> g = transform(g)
+
+        >>> # define and train the model
+        >>> model = Model(input_dim, hidden_dim, num_classes, g.canonical_etypes)
+        >>> optimizer = th.optim.Adam(model.parameters())
+        >>> for epoch in range(10):
+        ...     logits = model(g, g.ndata["h"])
+        ...     loss = th.nn.functional.cross_entropy(logits, th.tensor([1]))
+        ...     optimizer.zero_grad()
+        ...     loss.backward()
+        ...     optimizer.step()
+
+        >>> # Initialize the explainer
+        >>> explainer = dgl.nn.HeteroPGExplainer(model, hidden_dim)
+
+        >>> # Train the explainer
+        >>> # Define explainer temperature parameter
+        >>> init_tmp, final_tmp = 5.0, 1.0
+        >>> optimizer_exp = th.optim.Adam(explainer.parameters(), lr=0.01)
+        >>> for epoch in range(20):
+        ...     tmp = float(init_tmp * np.power(final_tmp / init_tmp, epoch / 20))
+        ...     loss = explainer.train_step(g, g.ndata["h"], tmp)
+        ...     optimizer_exp.zero_grad()
+        ...     loss.backward()
+        ...     optimizer_exp.step()
+
+        >>> # Explain the graph
+        >>> feat = g.ndata.pop("h")
+        >>> probs, edge_mask = explainer.explain_graph(g, feat)
```

Comment:
One last comment. I got 

```
{('game',
  'rev_plays',
  'user'): tensor([3.0763e-04, 1.3422e-04, 9.6319e-07, 2.5353e-05],
        grad_fn=<IndexBackward0>),
 ('user',
  'plays',
  'game'): tensor([3.0763e-04, 1.3422e-04, 9.6319e-07, 2.5353e-05],
        grad_fn=<IndexBackward0>)}
```

for `edge_mask`. In other words, the reverse edges have the same importance as the forward edges. Is this expected from the perspective of `PGExplainer`?

Reply:
`
To explain predictions made by a GNN model, we divide the original input graph Go into two subgraphs: Go = Gs + ∆G, where Gs presents the underlying subgraph that makes important contributions to GNN’s predictions, which is the expected explanatory graph, and ∆G consists of the remaining task-irrelevant edges for predictions made by the GNN
`

From the above line in the paper, I understand that when we select an edge as important (e.g., be part of `Gs`), we also select its reverse edge.  Therefore, the reverse edges are expected to have the same importance as the forward edges of the selected edges.

Reply:
let me review the reference material for any other possible remarks

Reply:
code-wise this is a result of the following logic within the `explain_graph` call, which obtains symmetric edge weights:

```python
reverse_eids = homo_graph.edge_ids(row, col).long()
edge_mask = (values + values[reverse_eids]) / 2
```

If I had to rationalize this, I would say it is due to the implicit requirement that every node has a corresponding embedding, which explains the addition of reverse-edges in order to guarantee global message-passing coverage. Then as a result we assume that these edges share the same score?

Reply:
@ndbaker1 @kunmukh Is this from a prior implementation? I did not find this in the paper.

Reply:
In the original implementation, https://github.com/flyingdoog/PGExplainer/blob/master/codes/Explainer.py#L27, I saw this line `sym_mask = (sym_mask + tf.transpose(sym_mask)) / 2`, where they calculate a symmetric mask by adding sym_mask with its transpose and dividing the result by 2. This ensures that the mask is symmetric along the diagonal.

Since the mask is symmetric along the diagonal, the edges and reverse edges have `edge_mask = (values + values[reverse_eids]) / 2` . 

@ndbaker1, please correct me if I missed something.

Reply:
I see. Thanks. I'm good then.